nautobot/extras/datasources/git.py

```diff
@@ -935,6 +936,124 @@ def delete_git_export_templates(repository_record, job_result, preserve=None):
             job_result.log(msg, level_choice=LogLevelChoices.LOG_WARNING, grouping="export templates")
 
 
+#
+# GraphQL handling
+#
+
+
+def refresh_git_graphql_queries(repository_record, job_result, delete=False):
+    """Callback function for GitRepository updates - refresh all GraphQLQuery managed by this repository."""
+    if "extras.graphqlquery" in repository_record.provided_contents and not delete:
+        update_git_graphql_queries(repository_record, job_result)
+    else:
+        delete_git_graphql_queries(repository_record, job_result)
+
+
+logger = logging.getLogger(__name__)
+
+
+def update_git_graphql_queries(repository_record, job_result):
+    """Refresh any GraphQL queries provided by this Git repository."""
+    graphql_query_path = os.path.join(repository_record.filesystem_path, "graphql_queries")
+    git_repository_content_type = ContentType.objects.get_for_model(GitRepository)
+    graphql_queries = []
+
+    if os.path.isdir(graphql_query_path):
+        for file in os.listdir(graphql_query_path):
+            file_path = os.path.join(graphql_query_path, file)
+            if not os.path.isfile(file_path):
+                continue
+
+            # Remove `.gql` extension from the name if it exists
+            query_name = file.rsplit(".gql", 1)[0] if file.endswith(".gql") else file
+
+            try:
+                with open(file_path, "r") as fd:
+                    query_content = fd.read().strip()
+
+                graphql_query, created = GraphQLQuery.objects.get_or_create(
+                    name=query_name,
+                    owner_content_type=git_repository_content_type,
+                    owner_object_id=repository_record.pk,
+                    defaults={"query": query_content},
+                )
+                modified = graphql_query.query != query_content
+
+                # Only attempt to update if the content has changed
+                if modified:
+                    graphql_query.query = query_content
+                    try:
+                        # Validate and save the query
+                        graphql_query.full_clean()
+                        graphql_query.save()
+                        msg = (
+                            f"Successfully created GraphQL query: {query_name}"
+                            if created
+                            else f"Successfully updated GraphQL query: {query_name}"
+                        )
+                        graphql_queries.append(query_name)
```

Comment:
Could move this assignment out of the various if/else try/except cases and just do it once at line 981.